gtsfm/utils/features.py

```diff
@@ -45,5 +45,67 @@ def normalize_coordinates(coordinates: np.ndarray, intrinsics: Cal3Bundler) -> n
     Returns:
         normalized coordinates, of shape Nx2.
     """
-
     return np.vstack([intrinsics.calibrate(x[:2].reshape(2, 1)) for x in coordinates])
+
+
+def convert_to_homogenous_coordinates(
+    non_homogenous_coordinates: np.ndarray,
+) -> np.ndarray:
+    """Convert coordinates to homogenous system (by appending a column of ones).
+
+    Args:
+        non_homogenous_coordinates: 2d non-homogenous coordinates, of shape Nx2.
+
+    Returns:
+        2d homogenous coordinates, of shape Nx3.
+
+    Raises:
+        TypeError: if input is not 2 dimensional.
+    """
+    if non_homogenous_coordinates is None or non_homogenous_coordinates.size == 0:
+        return non_homogenous_coordinates
+
+    if non_homogenous_coordinates.shape[1] != 2:
+        raise TypeError("Input should be 2D")
+
+    return np.hstack(
+        (
+            non_homogenous_coordinates,
+            np.ones((non_homogenous_coordinates.shape[0], 1)),
+        )
+    )
+
+
+def convert_to_epipolar_lines(normalized_coordinates_i1: np.ndarray, i2Ei1: EssentialMatrix) -> np.array:
+    """Convert coordinates to epipolar lines in image i2.
+
+    The epipolar line in image i2 is given by i2Ei1 @ x_i1. A point x_i2 is on this line if x_i2^T @ i2Ei1 @ x_i1 = 0.
+
+    Args:
+        normalized_coordinates_i1: normalized coordinates in i1, of shape Nx2.
+        i2Ei1: essential matrix.
+
+    Returns:
+        Corr. epipolar lines in i2, of shape Nx3
+    """
+    if normalized_coordinates_i1 is None or normalized_coordinates_i1.size == 0:
+        return normalized_coordinates_i1
```

Comment:
are we handling this failure mode correctly? why return normalized coords instead of epipolar lines? should we return None?

Reply:
I wanted to return an empty array if the input was empty, and None if the input was None. I thought this will be easier to do instead of two if conditions

Reply:
aren't we returning an incorrect quantity though?

Reply:
I did not get this.

Reply:
@ayushbaid I just meant that this function is supposed to return epipolar_lines.

But if we `return normalized_coordinates_i1`, we are not returning what this function is supposed to return. So we can't trust its output in that case, and we'll get inaccurate distances down the road.

I would prefer we return something like None and catch it, so we can abort for this image pair.

Reply:
got it. Makes sense and I'll implement this. I guess returning None is cleaner so I'll make that change.